src/test/kotlin/cacophony/automata/DFAEquivalenceTest.kt

```diff
@@ -0,0 +1,393 @@
+package cacophony.automata
+
+import org.junit.jupiter.api.Assertions.assertFalse
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.Test
+
+class DFAEquivalenceTest {
+    private fun createDFA(
+        starting: Int,
+        productions: Map<Pair<Int, Char>, Int>,
+        accepting: Set<Int>,
+    ): DFA<Int> {
+        val allStates = (setOf(starting) union accepting union productions.keys.map { it.first }.toSet() union productions.values).toList()
+        return object : DFA<Int> {
+            override fun getStartingState(): Int {
+                return starting
+            }
+
+            override fun getAllStates(): List<Int> {
+                return allStates
+            }
+
+            override fun getProductions(): Map<Pair<Int, Char>, Int> {
+                return productions
+            }
+
+            override fun getProduction(
+                state: Int,
+                symbol: Char,
+            ): Int? {
+                return productions[Pair(state, symbol)]
+            }
+
+            override fun isAccepting(state: Int): Boolean {
+                return accepting.contains(state)
+            }
+        }
+    }
+
+    @Test
+    fun testDFAIsEquivalentToItself() {
+        val dfa =
+            createDFA(
+                42,
+                mapOf(
+                    Pair(42, 'b') to 43,
+                    Pair(43, 'c') to 53,
+                ),
+                setOf(53),
+            )
+        assertTrue(areEquivalent(dfa, dfa))
+    }
+
+    @Test
+    fun testTwoDFAsForTheSameSimpleLanguageAndOfTheSameSizeAreEquivalent() {
+        // These two DFAs accept all words that contain at least one 'b'
+        // They are identical up to state renaming
+        val first =
+            createDFA(
+                1,
+                mapOf(
+                    Pair(1, 'a') to 1,
+                    Pair(1, 'b') to 2,
+                    Pair(2, 'a') to 2,
+                    Pair(2, 'b') to 2,
+                ),
+                setOf(2),
+            )
+        val second =
+            createDFA(
+                2,
+                mapOf(
+                    Pair(2, 'a') to 2,
+                    Pair(2, 'b') to 1,
+                    Pair(1, 'a') to 1,
+                    Pair(1, 'b') to 1,
+                ),
+                setOf(1),
+            )
+        assertTrue(areEquivalent(first, second))
+    }
+
+    @Test
+    fun testTwoDFAsForDifferentSimpleLanguagesButOfTheSameSizeAreNotEquivalent() {
```

Comment:
please use `snake_case` for test methods, especially when they are this long

Reply:
As discussed, I switched to using the Kotlin feature where one can enclose the test name in backticks